contracts/governance/proposals.cairo

```diff
@@ -0,0 +1,324 @@
+%lang starknet
+
+from starkware.starknet.common.syscalls import get_block_number, get_caller_address
+from starkware.cairo.common.uint256 import Uint256, uint256_mul, assert_uint256_lt, uint256_lt
+from starkware.cairo.common.math import assert_not_zero, assert_le, assert_nn, unsigned_div_rem, assert_nn_le
+from starkware.cairo.common.math_cmp import is_le, is_nn
+
+from types import Address, PropDetails, BlockNumber, VoteStatus, ContractType, VoteCounts
+from gov_constants import PROPOSAL_VOTING_TIME_BLOCKS, NEW_PROPOSAL_QUORUM, QUORUM, TEAM_TOKEN_BALANCE
+from gov_helpers import intToUint256
+
+from openzeppelin.token.erc20.IERC20 import IERC20
+
+@event
+func Proposed(prop_id: felt, impl_hash: felt, to_upgrade: ContractType) {
+}
+
+@event
+func Voted(prop_id: felt, voter: Address, opinion: VoteStatus) {
+}
+
+@storage_var
+func proposal_details(prop_id: felt) -> (res: PropDetails) {
+}
+
+@storage_var
+func proposal_vote_ends(prop_id: felt) -> (block_number: BlockNumber) {
+}
+
+// 0 = not voted, 1 = yay, -1 = nay
+@storage_var
+func proposal_voted_by(prop_id: felt, token_holder: Address) -> (res: VoteStatus) {
```

Comment:
Do we need to store this? Seems redundant since it's all on chain anyway

Reply:
We need it to prevent an account from voting twice